File: server/src/models/Student.js
const mongoose = require('mongoose');

const studentSchema = new mongoose.Schema({
    studentId: {
        type: String,
        required: true,
        unique: true,
        index: true
    },
    name: {
        first: { type: String, required: true },
        last: { type: String, required: true }
    },
    dob: {
        type: Date,
        required: true
    },
    gender: {
        type: String,
        enum: ['M', 'F', 'Other'],
        required: true
    },
    class: {
        type: String,
        required: true
    },
    section: {
        type: String,
        required: true
    },
    guardians: [{
        name: { type: String, required: true },
        relation: { type: String, required: true },
        phone: { type: String, required: true },
        email: { type: String }
    }],
    medical: {
        allergies: [String],
        notes: String
    },
    photoUrl: String,
    status: {
        type: String,
        enum: ['active', 'inactive', 'graduated', 'transferred'],
        default: 'active'
    },
    firebaseUid: {
        type: String,
        sparse: true,
        index: true
    },
    admissionDate: {
        type: Date,
        default: Date.now
    }
}, {
    timestamps: true
});

// Virtual for full name
studentSchema.virtual('fullName').get(function () {
    return `${this.name.first} ${this.name.last}`;
});

// Virtual for age
studentSchema.virtual('age').get(function () {
    const today = new Date();
    const birthDate = new Date(this.dob);
    let age = today.getFullYear() - birthDate.getFullYear();
    const monthDiff = today.getMonth() - birthDate.getMonth();
    if (monthDiff < 0 || (monthDiff === 0 && today.getDate() < birthDate.getDate())) {
        age--;
    }
    return age;
});

// Index for searching
studentSchema.index({ 'name.first': 'text', 'name.last': 'text', studentId: 'text' });

studentSchema.set('toJSON', { virtuals: true });
studentSchema.set('toObject', { virtuals: true });

module.exports = mongoose.model('Student', studentSchema);

File: mobile/lib/src/data/models/student.dart
class Student {
  final String id;
  final String studentId;
  final String firstName;
  final String lastName;
  final DateTime dob;
  final String gender;
  final String className;
  final String section;
  final List<Guardian> guardians;
  final Medical medical;
  final String? photoUrl;
  final String status;
  final String? firebaseUid;
  final DateTime admissionDate;

  Student({
    required this.id,
    required this.studentId,
    required this.firstName,
    required this.lastName,
    required this.dob,
    required this.gender,
    required this.className,
    required this.section,
    required this.guardians,
    required this.medical,
    this.photoUrl,
    required this.status,
    this.firebaseUid,
    required this.admissionDate,
  });

  String get fullName => '$firstName $lastName';

  factory Student.fromJson(Map<String, dynamic> json) {
    return Student(
      id: json['_id'] ?? json['id'],
      studentId: json['studentId'],
      firstName: json['name']['first'],
      lastName: json['name']['last'],
      dob: DateTime.parse(json['dob']),
      gender: json['gender'],
      className: json['class'],
      section: json['section'],
      guardians: (json['guardians'] as List)
          .map((g) => Guardian.fromJson(g))
          .toList(),
      medical: Medical.fromJson(json['medical'] ?? {}),
      photoUrl: json['photoUrl'],
      status: json['status'],
      firebaseUid: json['firebaseUid'],
      admissionDate: DateTime.parse(json['admissionDate']),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'studentId': studentId,
      'name': {'first': firstName, 'last': lastName},
      'dob': dob.toIso8601String(),
      'gender': gender,
      'class': className,
      'section': section,
      'guardians': guardians.map((g) => g.toJson()).toList(),
      'medical': medical.toJson(),
      'photoUrl': photoUrl,
      'status': status,
      'firebaseUid': firebaseUid,
    };
  }
}

class Guardian {
  final String name;
  final String relation;
  final String phone;
  final String? email;

  Guardian({
    required this.name,
    required this.relation,
    required this.phone,
    this.email,
  });

  factory Guardian.fromJson(Map<String, dynamic> json) {
    return Guardian(
      name: json['name'],
      relation: json['relation'],
      phone: json['phone'],
      email: json['email'],
    );
  }

  Map<String, dynamic> toJson() {
    return {'name': name, 'relation': relation, 'phone': phone, 'email': email};
  }
}

class Medical {
  final List<String> allergies;
  final String? notes;

  Medical({this.allergies = const [], this.notes});

  factory Medical.fromJson(Map<String, dynamic> json) {
    return Medical(
      allergies: json['allergies'] != null
          ? List<String>.from(json['allergies'])
          : [],
      notes: json['notes'],
    );
  }

  Map<String, dynamic> toJson() {
    return {'allergies': allergies, 'notes': notes};
  }
}
